File: src/backend/TrafficCourts/Citizen.Service/Models/Tickets/ViolationTicketCount.cs
using System.ComponentModel.DataAnnotations;
using System.Text.Json.Serialization;
using TrafficCourts.Common.OpenAPIs.OracleDataApi.v1_0;

namespace TrafficCourts.Citizen.Service.Models.Tickets;

/// <summary>
/// Represents a violation ticket count.
/// </summary>
public class ViolationTicketCount
{
    /// <summary>
    /// The count number. Must be unique within an individual violation ticket.
    /// </summary>
    [JsonPropertyName("count_no")]
    [Range(1, 3)]
    public short CountNo { get; set; }    

    /// <summary>
    /// The description of the offence.
    /// </summary>
    [JsonPropertyName("description")]
    [MaxLength(50)]
    public string? Description { get; set; }

    /// <summary>
    /// The act or regulation code the violation occurred against. For example, MVA, WLA, TCR, etc
    /// </summary>
    [JsonPropertyName("act_or_regulation_name_code")]
    [MaxLength(3)]
    public string? ActOrRegulationNameCode { get; set; }

    /// <summary>
    /// The full section designation of the act or regulation. For example, "147(1)" which means "Speed in school zone"
    /// </summary>
    [JsonPropertyName("full_section")]
    [MaxLength(20)]
    public string? FullSection { get; set; }

    /// <summary>
    /// The ticketed amount.
    /// </summary>
    /// <remarks>The upper bound arbitrarily set to just under one million.</remarks>
    [JsonPropertyName("ticketed_amount")]
    [Range(1f, 999_999f)]
    public decimal? TicketedAmount { get; set; }

    /// <summary>
    /// The count is flagged as an offence to an act. Cannot be true, if is_regulation is true.
    /// </summary>
    [JsonPropertyName("is_act")]
    public ViolationTicketCountIsAct? IsAct { get; set; }

    /// <summary>
    /// The count is flagged as an offence to a regulation. Cannot be true, if is_act is true.
    /// </summary>
    [JsonPropertyName("is_regulation")]
    public ViolationTicketCountIsRegulation? IsRegulation { get; set; }
}

File: src/backend/oracle-data-api/src/main/java/ca/bc/gov/open/jag/tco/oracledataapi/model/ViolationTicketCount.java
package ca.bc.gov.open.jag.tco.oracledataapi.model;

import jakarta.validation.constraints.Max;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.Size;

import com.fasterxml.jackson.annotation.JsonBackReference;

import io.swagger.v3.oas.annotations.media.Schema;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;


/**
 * @author 237563
 * Represents a violation ticket count.
 *
 */

@Getter
@Setter
@NoArgsConstructor
public class ViolationTicketCount extends Auditable<String> {

	@Schema(description = "ID", accessMode = Schema.AccessMode.READ_ONLY)
	private Long violationTicketCountId;

	/**
	 * The count number.
	 */
	@Min(1)
	@Max(3)
	private int countNo;

	/**
	 * The description of the offence.
	 */
	@Schema(nullable = true)
	private String description;

	/**
	 * The act or regulation code the violation occurred against. For example, MVA, WLA, TCR, etc
	 */
	@Schema(nullable = true)
	private String actOrRegulationNameCode;

	/**
	 * The count is flagged as an offence to an act. Cannot be true, if is_regulation is true.
	 */
	@Schema(nullable = true)
	private YesNo isAct;

	/**
	 * The count is flagged as an offence to a regulation. Cannot be true, if is_act is true.
	 */
	@Schema(nullable = true)
	private YesNo isRegulation;

	/**
	 * The section part of the full section. For example, "127"
	 */
	@Size(max = 10)
	@Schema(nullable = true, maxLength = 10, accessMode = Schema.AccessMode.READ_ONLY)
	private String section;

	/**
	 * The subsection part of the full section. For example, "(1)"
	 */
	@Size(max = 4)
	@Schema(nullable = true, maxLength = 4, accessMode = Schema.AccessMode.READ_ONLY)
	private String subsection;

	/**
	 * The paragraph part of the full section. For example, "(a)"
	 */
	@Size(max = 3)
	@Schema(nullable = true, maxLength = 3, accessMode = Schema.AccessMode.READ_ONLY)
	private String paragraph;

	/**
	 * The subparagraph part of the full section. For example, "(ii)"
	 */
	@Size(max = 5)
	@Schema(nullable = true, maxLength = 5, accessMode = Schema.AccessMode.READ_ONLY)
	private String subparagraph;

	/**
	 * The ticketed amount.
	 */
	@Schema(nullable = true)
	private Float ticketedAmount;

	@JsonBackReference
	@Schema(hidden = true)
	private ViolationTicket violationTicket;

}
